src/PostgREST/Config.hs

```diff
@@ -371,37 +358,65 @@ readAppConfig dbSettings env optPath dbUriFile secretFile = do
+    readSecret filename = do
+      jwtSecret <- chomp <$> BS.readFile (toS filename)
+      return $ conf { configJwtSecret = Just jwtSecret }
+    chomp bs = fromMaybe bs (BS.stripSuffix "\n" bs)
+
+decodeSecret :: AppConfig -> IO AppConfig
+decodeSecret conf@AppConfig{..} =
+  case (configJwtSecretIsBase64, configJwtSecret) of
+    (True, Just secret) ->
+      either fail (return . updateSecret) $ decodeB64 secret
+    _ -> return conf
+  where
+    updateSecret bs = conf { configJwtSecret = Just bs }
+    decodeB64 = B64.decode . encodeUtf8 . T.strip . replaceUrlChars . decodeUtf8
+    replaceUrlChars = T.replace "_" "/" . T.replace "-" "+" . T.replace "." "="
+
+-- | Parse `jwt-secret` configuration option and turn into a JWKSet.
+--
+-- There are three ways to specify `jwt-secret`: text secret, JSON Web Key
+-- (JWK), or JSON Web Key Set (JWKS). The first two are converted into a JWKSet
+-- with one key and the last is converted as is.
+decodeJWKS :: AppConfig -> AppConfig
+decodeJWKS conf =
+  conf { configJWKS = parseSecret <$> configJwtSecret conf }
 
-  There are three ways to specify `jwt-secret`: text secret, JSON Web Key
-  (JWK), or JSON Web Key Set (JWKS). The first two are converted into a JWKSet
-  with one key and the last is converted as is.
--}
 parseSecret :: ByteString -> JWKSet
 parseSecret bytes =
   fromMaybe (maybe secret (\jwk' -> JWT.JWKSet [jwk']) maybeJWK)
     maybeJWKSet
- where
-  maybeJWKSet = JSON.decode (toS bytes) :: Maybe JWKSet
-  maybeJWK = JSON.decode (toS bytes) :: Maybe JWK
-  secret = JWT.JWKSet [JWT.fromKeyMaterial keyMaterial]
-  keyMaterial = JWT.OctKeyMaterial . JWT.OctKeyParameters $ JOSE.Base64Octets bytes
-
--- | Read the JWT secret from a file if configJwtSecret is actually a filepath(has @ as its prefix).
--- | To check if the JWT secret is provided is in fact a file path, it must be decoded as 'Text' to be processed.
-readSecretFile :: Maybe B.ByteString -> IO (Maybe B.ByteString)
-readSecretFile mSecret =
-  case (T.stripPrefix "@" . decodeUtf8) =<< mSecret of
-    Nothing       -> return Nothing
-    Just filename -> Just . chomp <$> BS.readFile (toS filename)
   where
-    chomp bs = fromMaybe bs (BS.stripSuffix "\n" bs)
+    maybeJWKSet = JSON.decode (toS bytes) :: Maybe JWKSet
+    maybeJWK = JSON.decode (toS bytes) :: Maybe JWK
+    secret = JWT.JWKSet [JWT.fromKeyMaterial keyMaterial]
+    keyMaterial = JWT.OctKeyMaterial . JWT.OctKeyParameters $ JOSE.Base64Octets bytes
 
 -- | Read database uri from a separate file if `db-uri` is a filepath.
-readDbUriFile :: Text -> IO (Maybe Text)
-readDbUriFile dbUri = case T.stripPrefix "@" dbUri of
-  Nothing       -> return Nothing
-  Just filename -> Just . T.strip <$> readFile (toS filename)
+readDbUriFile :: Maybe AppConfig -> AppConfig -> IO AppConfig
+readDbUriFile maybePrevConf conf =
+  case maybePrevConf of
+    Just prevConf ->
+      pure $ conf { configDbUri = configDbUri prevConf }
```

Comment:
Can you expand on this a little bit? I think this is a change of behaviour - it's not possible to reload a different db-uri from the config file anymore, right? What is the background for this?

Reply:
Yes, good question! The behavior should be unchanged. The intention is to explicitly model what happens implicitly here: https://github.com/PostgREST/postgrest/blob/7ca0d469367d480804b2b8b91bdceb653ada754a/main/Main.hs#L146

Also note that the dbUri in the original code is only ever used [here on startup](https://github.com/PostgREST/postgrest/blob/4b46c4eff329c91dc13b26c78b266c7361ecace4/main/Main.hs#L125) and [here in the `listener`](https://github.com/PostgREST/postgrest/blob/4b46c4eff329c91dc13b26c78b266c7361ecace4/main/Main.hs#L359). It can never change once the dbUriFile has been read and Config parsed.

Reply:
I see. The reason why I looked at that so carefully, is that my passing in `previousConf` into `readAppConfig` that opens up all kind of doors where stuff could go wrong. E.g., if the full prev config was taken and new values applied on it - that would not allow resetting db-settings by just removing them and so on.

From the comment you linked, I understand that not reloading was the easy way out back then. But in our case it makes things more complicated - the easy way out after the refactor would be to just reload it anyway. We do have a few other settings that are not reloadable either (correct?), so after reloading those settings will not reflect reality anymore anyway - which is not a problem.

I suggest to just remove the whole `previousConf` stuff and reload it all the time. Makes it easier to understand, less-error prone for future "improvements", etc.

Reply:
Currently we can't re-read the dbUri, as it might be read from `stdin`, where it's only available once. I've restricted the arguments to `readAppConfig` for now, without changing the behavior

Reply:
A lot better, thanks! That makes it much easier to see what's happening.